<div style="text-align: right"> 
Universidad Nacional de Colombia <br>
Maestría en Ciencias - Estadística <br>
Big Data - Métodos Intensivos de Computación Estadística -2023829 <br>
Semestre 2021-1 
</div> 

<div style="text-align: right"> 
Estudiante: Diego Clemente Rojas Chingate   <br>
</div> 


# Dataset Reddit comments

Dataset de cada evento público de Github de los años 2014 y 2015 almacenado como "gzip" que contiene un Json delimitado por líneas (El esquema cambia entre 2014 y 2015). [Fuente](https://github.com/dask/dask-tutorial/blob/main/Homework.ipynb)

Realizar lo siguiente:

1. [Usar "dask.bag" para inspeccionar los datos.](#1.-Usar-"dask.bag"-para-inspeccionar-los-datos.)

2. [Profundizar en funciones como "pluck" y "filter".](#2.-Profundizar-en-funciones-como-"pluck"-y-"filter".)

3. [¿Quién fue el Commiter más popular de 2015?](#3.-¿Quién-fue-el-Commiter-más-popular-de-2015?)

# 1. Usar "dask.bag" para inspeccionar los datos.

### Expresión para obtener los archivos de las carpetas "./data/2014" y "./data/2015"

In [1]:
import os

data_2014 = os.path.join('./data/2014', '2014*.json.gz')
data_2015 = os.path.join('./data/2015', '0*/2015*.json.gz')

### Obtener los Bag de los archivos correspondientes a 2014 y 2015

In [2]:
import dask.bag as db

bag_data_2014 = db.read_text(data_2014)
bag_data_2015 = db.read_text(data_2015)


### Primeras 5 líneas (cada línea es un objeto JSON) de 2014

In [3]:
bag_data_2014.take(5)

('{"created_at":"2014-01-01T00:02:02-08:00","payload":{"shas":[["84d3daa33301d74a10cce1fc340e306b19108208","a4c90bd95206b987e4df905ee9f57ba9e064c73a@gmail.com","Clicking on link opens it in new tab now","Shrayas",true],["403f4884d3e1ae2651345304d989275c69eed65e","7b6523a745c0da3ed6e33c034f58cffde1f9de21@gmail.com","Merge pull request #28 from shrayas/open-link-in-new-tab\\n\\nClicking on link opens it in new tab now","Raghotham S",true]],"size":2,"ref":"refs/heads/master","head":"403f4884d3e1ae2651345304d989275c69eed65e"},"public":true,"type":"PushEvent","url":"https://github.com/raghothams/sharURL/compare/dea5599b6c...403f4884d3","actor":"raghothams","actor_attributes":{"login":"raghothams","type":"User","gravatar_id":"4a1a459a7121d36bbd0ad15b59735b50","name":"Raghotham S","company":"SAP Labs Pvt. Ltd.","location":"Bangalore","email":"7b6523a745c0da3ed6e33c034f58cffde1f9de21@gmail.com"},"repository":{"id":11090251,"name":"sharURL","url":"https://github.com/raghothams/sharURL","descrip

### Primeras 5 líneas (cada línea es un objeto JSON) de 2015

In [4]:
bag_data_2015.take(5)

('{"id":"2489368070","type":"PushEvent","actor":{"id":9152315,"login":"davidjhulse","gravatar_id":"","url":"https://api.github.com/users/davidjhulse","avatar_url":"https://avatars.githubusercontent.com/u/9152315?"},"repo":{"id":28635890,"name":"davidjhulse/davesbingrewardsbot","url":"https://api.github.com/repos/davidjhulse/davesbingrewardsbot"},"payload":{"push_id":536740396,"size":1,"distinct_size":1,"ref":"refs/heads/master","head":"a9b22a6d80c1e0bb49c1cf75a3c075b642c28f81","before":"86ffa724b4d70fce46e760f8cc080f5ec3d7d85f","commits":[{"sha":"a9b22a6d80c1e0bb49c1cf75a3c075b642c28f81","author":{"email":"da8d7d1118ca5befd4d0d3e4f449c76ba6f1ee7e@live.com","name":"davidjhulse"},"message":"Altered BingBot.jar\\n\\nFixed issue with multiple account support","distinct":true,"url":"https://api.github.com/repos/davidjhulse/davesbingrewardsbot/commits/a9b22a6d80c1e0bb49c1cf75a3c075b642c28f81"}]},"public":true,"created_at":"2015-01-01T00:00:00Z"}\n',
 '{"id":"2489368072","type":"PushEvent","a

### Mejorar la visualización de cada JSON en vez de mostrarlo como una línea 

In [5]:
import json
bag_data_2014_json = bag_data_2014.map(json.loads)
bag_data_2015_json = bag_data_2015.map(json.loads)

### Primera línea de 2014

In [6]:
bag_data_2014_json.take(1)

({'created_at': '2014-01-01T00:02:02-08:00',
  'payload': {'shas': [['84d3daa33301d74a10cce1fc340e306b19108208',
     'a4c90bd95206b987e4df905ee9f57ba9e064c73a@gmail.com',
     'Clicking on link opens it in new tab now',
     'Shrayas',
     True],
    ['403f4884d3e1ae2651345304d989275c69eed65e',
     '7b6523a745c0da3ed6e33c034f58cffde1f9de21@gmail.com',
     'Merge pull request #28 from shrayas/open-link-in-new-tab\n\nClicking on link opens it in new tab now',
     'Raghotham S',
     True]],
   'size': 2,
   'ref': 'refs/heads/master',
   'head': '403f4884d3e1ae2651345304d989275c69eed65e'},
  'public': True,
  'type': 'PushEvent',
  'url': 'https://github.com/raghothams/sharURL/compare/dea5599b6c...403f4884d3',
  'actor': 'raghothams',
  'actor_attributes': {'login': 'raghothams',
   'type': 'User',
   'gravatar_id': '4a1a459a7121d36bbd0ad15b59735b50',
   'name': 'Raghotham S',
   'company': 'SAP Labs Pvt. Ltd.',
   'location': 'Bangalore',
   'email': '7b6523a745c0da3ed6e33c034f58cf

### Primera línea de 2015

In [7]:
bag_data_2015_json.take(1)

({'id': '2489368070',
  'type': 'PushEvent',
  'actor': {'id': 9152315,
   'login': 'davidjhulse',
   'gravatar_id': '',
   'url': 'https://api.github.com/users/davidjhulse',
   'avatar_url': 'https://avatars.githubusercontent.com/u/9152315?'},
  'repo': {'id': 28635890,
   'name': 'davidjhulse/davesbingrewardsbot',
   'url': 'https://api.github.com/repos/davidjhulse/davesbingrewardsbot'},
  'payload': {'push_id': 536740396,
   'size': 1,
   'distinct_size': 1,
   'ref': 'refs/heads/master',
   'head': 'a9b22a6d80c1e0bb49c1cf75a3c075b642c28f81',
   'before': '86ffa724b4d70fce46e760f8cc080f5ec3d7d85f',
   'commits': [{'sha': 'a9b22a6d80c1e0bb49c1cf75a3c075b642c28f81',
     'author': {'email': 'da8d7d1118ca5befd4d0d3e4f449c76ba6f1ee7e@live.com',
      'name': 'davidjhulse'},
     'message': 'Altered BingBot.jar\n\nFixed issue with multiple account support',
     'distinct': True,
     'url': 'https://api.github.com/repos/davidjhulse/davesbingrewardsbot/commits/a9b22a6d80c1e0bb49c1cf75a3c

### ¿Cuántos eventos Github hubieron en 2014?

In [9]:
bag_data_2014_json.count().compute()

9001106

### ¿Cuántos eventos Github hubieron en 2015?

In [10]:
bag_data_2015_json.count().compute()

28506909

## 2. Profundizar en funciones como "pluck" y "filter".

### ¿Cuántos eventos del tipo 'PushEvent' hubieron en 2014?

In [11]:
bag_data_2014_json.filter(lambda record: record['type'] == 'PushEvent').count().compute()

4496746

### ¿Cuántos eventos del tipo 'PushEvent' hubieron en 2015?

In [12]:
bag_data_2015_json.filter(lambda record: record['type'] == 'PushEvent').count().compute()

14271557

### ¿Cuántos eventos Github corresponden al usuario 'raghothams' en 2014?

In [13]:
bag_data_2014_json.filter(lambda record: record['actor'] == 'raghothams').count().compute()

44

### ¿Cuántos eventos Github corresponden al usuario 'davidjhulse' en 2015?

In [14]:
bag_data_2015_json.filter(lambda record: record['actor']['login'] == 'davidjhulse').count().compute()

2

### ¿Qué repositorios usó el usuario 'terkel' en 2014?

In [15]:
bag_data_2014_json.filter(lambda record: record['actor'] == 'terkel').pluck('repository').pluck('name').distinct().compute()

['terkel.jp-stylesheets',
 'a-font-garde',
 'gh-test',
 'book-of-modern-frontend-tooling']

### ¿Qué repositorios usó el usuario 'jmoon018' en 2015?

In [16]:
bag_data_2015_json.filter(lambda record: record['actor']['login'] == 'jmoon018').pluck('repo').pluck('name').distinct().compute()

['jmoon018/rshell-unit-tester',
 'jmoon018/Pac-Vim',
 'jmoon018/PacVim',
 'jmoon018/jmoon018.github.io',
 'Keenan144/flash_cards',
 'wolfram74/world_full_of_squirrels',
 'jmoon018/mtd-f',
 'jmoon018/Squirrel-Cards',
 'jmoon018/TreeTraversal',
 'sf-squirrels-2015/ph2-pA-week-1-passion-project',
 'jmoon018/ph2-pA-week-1-passion-project',
 'Keenan144/Survey_Gorilla',
 'adellhk/ps_overflow',
 'jmoon018/decoupled_par_overflow']

## 3. ¿Quién fue el Commiter más popular de 2015?

In [52]:
from operator import add
def incr(tot, element):
    return element['commits'] + tot


commiter = bag_data_2015_json.filter(lambda record: record['type'] == 'PushEvent').map(lambda x: {"id": x["actor"]["id"], "name": x["actor"]["login"], "commits": len(x["payload"]["commits"])}).foldby(key='name', 
                   binop=incr, 
                   initial=0, 
                   combine=add, 
                   combine_initial=0).map(lambda x: (x[1], x[0])).topk(10).compute()
commiter

[(562157, 'mirror-updates'),
 (105482, 'KenanSulayman'),
 (74510, 'greatfire'),
 (52133, 'greatfirebot'),
 (40960, 'k-okada'),
 (40028, 'diversify-exp-user'),
 (36350, 'asfgit'),
 (32723, 'hubot'),
 (32536, 'meatballhat'),
 (31064, 'chapuni')]

### El commiter mas popular en 2015 fue: 'KenanSulayman' con '105482' commits